### PR TITLE
cherrypick: storage: overhaul GCQueue.shouldQueue

### PR DESCRIPTION
**Pull request by mberhault**
TL;DR: use a scale-invariant `shouldQueue` for the GC Queue to avoid it running
in a tight loop without actually deleting anything. Adjust this scalable score
so that GC is delayed when a large proportion of the data in the replica is
live. Slightly perturb returned scores to avoid groups of replicas becoming
eligible for GC at the same time repeatedly.

Closes #15897.

--

When a key of size `B` is deleted at timestamp `T` or superseded by a newer
version, it henceforth is accounted for in the range's `GCBytesAge`. At time
`S`, its contribution to age will be `B*seconds(S-T)`. The aggregate
`GCBytesAge` of all deleted versions in the cluster is what the GC queue at
the time of writing bases its `shouldQueue` method on.

If a replica is queued to have its old values garbage collected, its contents
are scanned. However, the values which are deleted follow a criterion that
isn't immediately connected to `GCBytesAge`: We (basically) delete everything
that's older than the Replica's `TTLSeconds`.

Thus, it's not obvious that garbage collection has the effect of reducing the
metric that we use to consider the replica for the next GC cycle, and it seems
that we messed it up.

The previous metric used for queueing: `GCBytesAge/(1<<20 * ttl)` does not
have the right scaling. For example, consider that a value of size `1mb` is
overwritten with a newer version. After `ttl` seconds, it contributes `1mb` to
`GCBytesAge`, and so the replica has a score of `1`, i.e. (roughly) the range
becomes interesting to the GC queue. When GC runs, it will delete value that
are `ttl` old, which our value is. But a Replica is ~64mb, so picture that
you have 64mb of key-value data all at the same timestamp, and they become
superseded. Already after `ttl/64`, the metric becomes 1, but they keys won't
be GC'able for another (63*ttl)/64. Thus, GC will run "all the time" long
before it can actually have an effect.

The metric with correct scaling must thus take into account the size of the
range. What size exactly? Any data that isn't live (i.e. isn't readable by
a scan from the far future). That's `KeyBytes + ms.ValBytes - ms.LiveBytes`,
which is also known as `GCBytes` in the code. Hence, the better metric is
`GCBytesAge/(ttl*GCBytes)`.

Using this metric guarantees that after truncation, `GCBytesAge` is at most
`ttl*GCBytes` (where `GCBytes` has been updated), i.e. the new metric is at
most 1.

To visualize this, picture a rectangular frame of width `ttl` and height
`GCBytes` (i.e. the horizontal dimension is time, the vertical one bytes),
where the right boundary of the frame corresponds to age zero.
Each non-live key is a domino aligned with the right side of the frame, its
width equal to its size, and its height given by the duration (in seconds) it's
been non-live.

The combined surface of the dominos is then `GCBytesAge`, and the claim is that
if the total sum of domino heights (i.e. sizes) is `GCBytes`, and the surface
is larger than `ttl*GCBytes` by some positive `X`, then after removing the
dominos that cross the line `x=-ttl` (i.e. `ttl` to the left from the right
side of the frame), at least a surface area of `X` has been removed.

```           x=-ttl                 GCBytes=1+4
                |           3 (age)
                |          +-------+
                |          | keep  | 1 (bytes)
                |          +-------+
           +-----------------------+
           |                       |
           |        remove         | 3 (bytes)
           |                       |
           +-----------------------+
                |   7 (age)
```

This is true because

```
deletable area  = total area       - nondeletable area
                = X + ttl*GCBytes  - nondeletable area
               >= X + ttl*GCBytes  - ttl*(bytes in nondeletable area)
                = X + ttl*(GCBytes - bytes in nondeletable area)
               >= X.
```

Or, in other words, you can only hope to put `ttl*GCBytes` of area in the
"safe" rectangle. Once you've done that, everything else you put is going to be
deleted.

This means that running GC will always result in a `GCBytesAge` of `<=
ttl*GCBytes`, and that a decent trigger for GC is a multiple of `ttl*GCBytes`.

Limitations:

- The intent portion of `shouldQueue` could be another time bomb if it fires
  continuously for some reason.

### CODE REVIEW

**Comment by mberhault** [comment]
CC: @cockroachdb/release

**Comment by tschottdorf** [comment]
:lgtm: (no changes necessary afaict).

---

Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/storage/gc_queue.go, line 69 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16006#-KkN6BX0kv7NdBzFtvdg:-KkN6BX0kv7NdBzFtvdh:b-nep61g) ([raw file](https://github.com/cockroachdb/cockroach/blob/465c7926b1f118699ee20d7c29a12ccb56c991d9/pkg/storage/gc_queue.go#L69)):*
> ```Go
> 	abortCacheAgeThreshold = 5 * base.DefaultHeartbeatInterval
> 
> 	// thresholds used by shouldQueue to decide whether to queue for GC based
> ```

Capitalization also missing on master, so let's leave it here.

---

*[pkg/storage/gc_queue.go, line 534 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16006#-KkN6UQkrB6hxV3TOUe8:-KkN6UQl7woJVlmJsuII:b-56ig8u) ([raw file](https://github.com/cockroachdb/cockroach/blob/465c7926b1f118699ee20d7c29a12ccb56c991d9/pkg/storage/gc_queue.go#L534)):*
> ```Go
> 	// Yes, comparing floats to 0 is bad. However, FinalScore is either exactly zero or
> 	// far away from zero.
> 	if r.FinalScore == 0 {
> ```

This one's actually a bug, but not a bad one (and also on master). Reminder to self: send a fix. Doesn't have to be 1.0.1.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16006#-:-KkN6ndZHhVhgw463p0w:bjuto7r)*
